flyteplugins/go/tasks/plugins/webapi/agent/plugin.go

```diff
@@ -348,82 +257,19 @@ func buildTaskExecutionMetadata(taskExecutionMetadata core.TaskExecutionMetadata
 	}
 }
 
-func getFinalTimeout(operation string, agent *Agent) config.Duration {
-	if t, exists := agent.Timeouts[operation]; exists {
-		return t
-	}
-
-	return agent.DefaultTimeout
-}
-
-func getFinalContext(ctx context.Context, operation string, agent *Agent) (context.Context, context.CancelFunc) {
-	timeout := getFinalTimeout(operation, agent).Duration
-	if timeout == 0 {
-		return ctx, func() {}
-	}
-
-	return context.WithTimeout(ctx, timeout)
-}
-
-func initializeAgentRegistry(cfg *Config, connectionCache map[*Agent]*grpc.ClientConn, getAgentMetadataClientFunc GetAgentMetadataClientFunc) (map[string]*Agent, error) {
-	agentRegistry := make(map[string]*Agent)
-	var agentDeployments []*Agent
-
-	// Ensure that the old configuration is backward compatible
-	for taskType, agentID := range cfg.AgentForTaskTypes {
-		agentRegistry[taskType] = cfg.Agents[agentID]
-	}
-
-	if len(cfg.DefaultAgent.Endpoint) != 0 {
-		agentDeployments = append(agentDeployments, &cfg.DefaultAgent)
-	}
-	agentDeployments = append(agentDeployments, maps.Values(cfg.Agents)...)
-	for _, agentDeployment := range agentDeployments {
-		client, err := getAgentMetadataClientFunc(context.Background(), agentDeployment, connectionCache)
-		if err != nil {
-			return nil, fmt.Errorf("failed to connect to agent [%v] with error: [%v]", agentDeployment, err)
-		}
-
-		finalCtx, cancel := getFinalContext(context.Background(), "ListAgents", agentDeployment)
-		defer cancel()
-
-		res, err := client.ListAgents(finalCtx, &admin.ListAgentsRequest{})
-		if err != nil {
-			grpcStatus, ok := status.FromError(err)
-			if grpcStatus.Code() == codes.Unimplemented {
-				// we should not panic here, as we want to continue to support old agent settings
-				logger.Infof(context.Background(), "list agent method not implemented for agent: [%v]", agentDeployment)
-				continue
-			}
-
-			if !ok {
-				return nil, fmt.Errorf("failed to list agent: [%v] with a non-gRPC error: [%v]", agentDeployment, err)
-			}
-
-			return nil, fmt.Errorf("failed to list agent: [%v] with error: [%v]", agentDeployment, err)
-		}
-
-		agents := res.GetAgents()
-		for _, agent := range agents {
-			supportedTaskTypes := agent.SupportedTaskTypes
-			for _, supportedTaskType := range supportedTaskTypes {
-				agentRegistry[supportedTaskType] = agentDeployment
-			}
-		}
-	}
-
-	return agentRegistry, nil
-}
-
 func newAgentPlugin() webapi.PluginEntry {
-	cfg := GetConfig()
-	connectionCache := make(map[*Agent]*grpc.ClientConn)
-	agentRegistry, err := initializeAgentRegistry(cfg, connectionCache, getAgentMetadataClientFunc)
+	cs, err := initializeClients(context.Background())
```

Comment:
Could you please help me understand what happens if one of the agent endpoints is not available when flytepropeller is booting up? Would it just crash?

Reply:
If so, it is a very strong assumption that I'm afraid we cannot take.

Reply:
Even the plugin loading is guarded so a single failing plugin would not impact propeller booting, it is still undesired that one failed agent endpoint (out of for example 50 healthy ones) would fail the whole plugin loading. I think some kind of late binding would be nicer.

Reply:
We have to wait for every endpoint to start; otherwise, FlytePropeller will keep crashing.

Reply:
Really nice suggestions, I will discuss with @pingsutw and reply the result here.
I think late binding or some kind of `endpoint` will be used mechanism is really helpful.
For example, in k8s, maybe `readiness probe ` can help use realize lazy binding?
I am not 100% familiar with how agent is deployed in flyte cluster, but yes, I will try to help.

Thank you for always providing great advices.

Reply:
Or maybe, only need 1 agent is connected is enough

Reply:
I think this plugin used to work in a late binding way, meaning the connection was not established until the very first RPC. In this way the propeller does not depend on agents in the wild when booting up.

> For example, in k8s, maybe readiness probe can help us realize lazy binding?

I think we cannot make any assumption how agents are deployed, and only interface they expose to the plugin is gRPC endpoints, so we can only make sense of out those.

Reply:
Will update the solution to your case with Kevin this week, thank you!

Reply:
@honnix, We will add a watcher to solve your case!
Thank you for the patience, we will mention you to see the new update, thank you very much.

Reply:
@Future-Outlier That is fantastic. Thank you!